packages/flutter/lib/src/material/navigation_rail.dart

```diff
@@ -0,0 +1,761 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:ui';
+
+import 'package:flutter/widgets.dart';
+
+import '../../scheduler.dart';
+
+import 'color_scheme.dart';
+import 'ink_well.dart';
+import 'material.dart';
+import 'material_localizations.dart';
+import 'navigation_rail_theme.dart';
+import 'theme.dart';
+import 'theme_data.dart';
+
+/// A material widget that is meant to be displayed at the left or right of an
+/// app to navigate between a small number of views, typically between three and
+/// five.
+///
+/// A navigation rail is usually used inside a [Row] of a [Scaffold] body.
+///
+/// The navigation rail is meant for layouts with wide viewports, such as a
+/// desktop web or tablet landscape layout. For smaller layouts, like mobile
+/// portrait, a [BottomNavigationBar] should be used instead. Adaptive layouts
+/// can build different instances of the [Scaffold] in order to have a
+/// navigation rail for more horizontal layouts and a bottom navigation bar
+/// for more vertical layouts.
+///
+/// {@tool dartpad --template=stateful_widget_material}
+///
+/// This example shows a [NavigationRail] used within a Scaffold with 3
+/// [NavigationRailDestination]s. The main content is separated by a divider
+/// (although elevation on the navigation rail can be used instead). The
+/// `_currentIndex` updates according to the `onDestinationSelected` callback.
+///
+/// ```dart
+/// int _currentIndex = 0;
+///
+///  @override
+///  Widget build(BuildContext context) {
+///    return Scaffold(
+///      body: Row(
+///        children: <Widget>[
+///          NavigationRail(
+///            currentIndex: _currentIndex,
+///            labelType: NavigationRailLabelType.selected,
+///            destinations: [
+///              NavigationRailDestination(
+///                icon: Icon(Icons.favorite_border),
+///                activeIcon: Icon(Icons.favorite),
+///                label: Text('First'),
+///              ),
+///              NavigationRailDestination(
+///                icon: Icon(Icons.bookmark_border),
+///                activeIcon: Icon(Icons.book),
+///                label: Text('Second'),
+///              ),
+///              NavigationRailDestination(
+///                icon: Icon(Icons.star_border),
+///                activeIcon: Icon(Icons.star),
+///                label: Text('Third'),
+///              ),
+///            ],
+///            onDestinationSelected: (int index) {
+///              setState(() {
+///                _currentIndex = index;
+///              });
+///            },
+///          ),
+///          VerticalDivider(thickness: 1, width: 1),
+///          Expanded(
```

Comment:
Could put a "main content" comment here to tie the example back to the explanatory text